test_workspace/Queue.bosatsu

```diff
@@ -0,0 +1,107 @@
+package Queue
+
+export [ Queue,
+  empty_Queue, fold_Queue, push, pop, pop_value, pop_Queue, reverse_Queue, eq_Queue,
+  queue_to_List,
+]
+
+struct Queue(front: List[a], back: List[a])
+
+empty_Queue: forall a. Queue[a] = Queue([], [])
+
+# convenient local alias
+empty = empty_Queue
+
+def queue_from_List(list: List[a]) -> Queue[a]:
+  Queue(list, [])
+
+def push(queue: Queue[a], item: a) -> Queue[a]:
+  Queue(f, b) = queue
+  Queue(f, [item, *b])
+
+def pop(queue: Queue[a]) -> Option[(a, Queue[a])]:
+  match queue:
+    Queue([h, *t], b): Some((h, Queue(t, b)))
+    Queue([], b):
+      match b.reverse:
+        []: None
+        [h, *t]: Some((h, Queue(t, [])))
+
+def pop_value(queue: Queue[a]) -> Option[a]:
+  match pop(queue):
+    Some((a, _)): Some(a)
+    None: None
+
+# drop an item off and return the rest, or empty
+def pop_Queue(queue: Queue[a]) -> Queue[a]:
+  match pop(queue):
+    Some((_, queue)): queue
+    None: empty
+
+def fold_Queue(queue: Queue[a], init: b, fn: b -> a -> b) -> b:
+  Queue(f, b) = queue
+  front = f.foldLeft(init, fn)
+  b.reverse.foldLeft(front, fn)
+
+def reverse_Queue(queue: Queue[a]) -> Queue[a]:
+  Queue(f, b) = queue
+  Queue(b.reverse, f.reverse)
+
+def eq_Queue(eq_fn: a -> a -> Bool, left: Queue[a], right: Queue[a]) -> Bool:
+  (res, _) = left.fold_Queue((True, right), \(g, right), al ->
+    match g:
+      False: (False, empty)
+      True:
+        match pop(right):
+          None: (False, empty)
+          Some((ar, right)):
+            (eq_fn(al, ar), right)
+    )
+  res
+
+def queue_to_List(q: Queue[a]) -> List[a]:
+  Queue(f, b) = q
+  f.concat(b.reverse)
+
+########
+## Tests below
+########
+
+def eq_Opt(fn, a, b):
+  match (a, b):
+    (None, None): True
+    (None, Some(_)): False
+    (Some(_), None): False
+    (Some(a), Some(b)): fn(a, b)
```

Comment:
I find this a little verbose... I suspect it would be a little more clear if you did
```
def eq_Opt(fn, a, b):
  match (a, b):
    (None, None): True
    (Some(a), Some(b)): fn(a, b)
    _: False
```

Reply:
(oh, this is just for testing)